model/PermissionsService.php

```diff
@@ -178,14 +262,19 @@ private function deduplicateActions(array $actions): array
         return $actions;
     }
 
-    private function getResourcesToUpdate(
-        core_kernel_classes_Resource $resource,
-        bool $isRecursive
-    ): array {
+    /**
+     * Provides an array holding the provided resource and, if it is a class, all
+     * resources that are instances of the provided class or any of its descendants
+     * plus all descendant classes.
+     *
+     * @return core_kernel_classes_Resource[]
+     */
+    private function getResourcesByClassRecursive(core_kernel_classes_Resource $resource): array
```

Comment:
This is not recursive. It is only getting nested class `resources`, but not `sub classes`. I think we should change the name.

Reply:
Checking: I think this should return both class and non-class resources (that is, all the resource tree using `$resource` as the root)

Iit is used from the code providing compatibility for legacy calls, and it is only called if `$command->isRecursive()==true`. We are not passing commands with the recursive flag set anymore unless they come from a call to one of the deprecated methods.

Reply:
I'll recheck locally if such calls are still working as before.

Reply:
I'm investigating this: Recursive calls are changing the permissions for nested classes, but not for nested resources.

Reply:
There was a bug indeed, fixed in a8b3b6dc7a58782a94a932050200c1a4c79ebc49, thanks :slightly_smiling_face: